File: TabCommon/Library/src/main/java/com/nextzy/library/base/mvvm/layer1View/BaseMvvmListAdapter.java
package com.nextzy.library.base.mvvm.layer1View;


import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import com.nextzy.library.base.mvvm.exception.ViewModelNotNullException;
import com.nextzy.library.base.mvvm.exception.ViewModelNotSetupException;
import com.nextzy.library.base.mvvm.layer2ViewModel.BaseViewModel;
import com.nextzy.library.base.view.holder.base.BaseViewHolder;

import java.lang.ref.WeakReference;


/**
 * Created by「 The Khaeng 」on 20 Sep 2017 :)
 */

public abstract class BaseMvvmListAdapter<VH extends BaseViewHolder, VM extends BaseViewModel>
        extends RecyclerView.Adapter<VH>{

    protected final Context context;
    protected WeakReference<FragmentActivity> activity;
    protected WeakReference<Fragment> fragment;
    protected OnClickHolderItemListener<VH> listener;

    public interface OnClickHolderItemListener<VH>{
        void onClickHolder( VH v, Object item, int position );
    }

    private RecyclerView recyclerView;

    public BaseMvvmListAdapter( FragmentActivity activity ){
        this.context = activity;
        this.activity = new WeakReference<>( activity );
    }

    public BaseMvvmListAdapter( Fragment fragment ){
        this.context = fragment.getContext();
        this.fragment = new WeakReference<>( fragment );
    }


    public void setOnClickHolderItemListener( OnClickHolderItemListener<VH> listener ){
        this.listener = listener;
    }


    @Override
    public void onAttachedToRecyclerView( RecyclerView recyclerView ){
        super.onAttachedToRecyclerView( recyclerView );
        this.recyclerView = recyclerView;
    }

    public RecyclerView getRecyclerView(){
        return recyclerView;
    }

    public VM getViewModelShared(){
        if( setupViewModel() == null ) throw new ViewModelNotSetupException();
        if( fragment != null ){
            return ViewModelProviders.of( fragment.get().getActivity() )
                    .get( setupViewModel() );
        }else if( activity != null ){
            return ViewModelProviders.of( activity.get() )
                    .get( setupViewModel() );
        }
        return null;
    }

    public VM getViewModel(){
        if( setupViewModel() == null ) throw new ViewModelNotSetupException();
        if( fragment != null ){
            return ViewModelProviders.of( fragment.get() )
                    .get( setupViewModel() );
        }else if( activity != null ){
            return ViewModelProviders.of( activity.get() )
                    .get( setupViewModel() );
        }
        return null;
    }

    public VM getViewModel( Class<VM> viewModelClass ){
        if( setupViewModel() == null ) throw new ViewModelNotNullException();
        if( fragment != null ){
            return ViewModelProviders.of( fragment.get() )
                    .get( viewModelClass );
        }else if( activity != null ){
            return ViewModelProviders.of( activity.get() )
                    .get( viewModelClass );
        }
        return null;
    }

    public abstract Class<VM> setupViewModel();


    @Override
    public VH onCreateViewHolder( ViewGroup parent, int viewType ){
        return null;
    }

    @Override
    public void onBindViewHolder( VH holder, int position ){
        if( listener != null ){
            holder.itemView.getRootView().setOnClickListener( setOnClickItem( holder ) );
        }
    }


    @NonNull
    private View.OnClickListener setOnClickItem( final VH holder ){
        return v -> listener.onClickHolder(
                holder,
                holder.getItem(),
                holder.getAdapterPosition()
        );
    }

}

File: TabCommon/Library/src/main/java/com/nextzy/library/base/mvvm/layer1View/BaseMvvmListAdapter.kt
package com.nextzy.library.base.mvvm.layer1View


import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import com.nextzy.library.base.mvvm.exception.ViewModelNotNullException
import com.nextzy.library.base.mvvm.exception.ViewModelNotSetupException
import com.nextzy.library.base.mvvm.layer2ViewModel.BaseViewModel
import com.nextzy.library.base.view.holder.base.BaseViewHolder
import java.lang.ref.WeakReference


/**
 * Created by「 The Khaeng 」on 20 Sep 2017 :)
 */

abstract class BaseMvvmListAdapter<VH : BaseViewHolder<*>, VM : BaseViewModel> : RecyclerView.Adapter<VH> {

    protected val context: Context
    protected var activity: WeakReference<FragmentActivity>? = null
    protected var fragment: WeakReference<Fragment>? = null
    protected var listener: OnClickHolderItemListener<VH>? = null

    var recyclerView: RecyclerView? = null
        private set

    val viewModelShared: VM?
        get() {
            if (setupViewModel() == null) throw ViewModelNotSetupException()
            if (fragment != null) {
                return ViewModelProviders.of(fragment?.get()?.activity!!)
                        .get(setupViewModel())
            } else if (activity != null) {
                return ViewModelProviders.of(activity?.get()!!)
                        .get(setupViewModel())
            }
            return null
        }

    val viewModel: VM
        get() {
            if (setupViewModel() == null) throw ViewModelNotSetupException()
            return if (fragment != null) {
                ViewModelProviders.of(fragment?.get()!!)
                        .get(setupViewModel())
            } else {
                ViewModelProviders.of(activity?.get()!!)
                        .get(setupViewModel())
            }
        }

    interface OnClickHolderItemListener<in VH> {
        fun onClickHolder(v: VH, position: Int)
    }

    constructor(activity: FragmentActivity) {
        this.context = activity
        this.activity = WeakReference(activity)
    }

    constructor(fragment: Fragment) {
        this.context = fragment.context
        this.fragment = WeakReference(fragment)
    }


    fun setOnClickHolderItemListener(listener: OnClickHolderItemListener<VH>) {
        this.listener = listener
    }


    override
    fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        this.recyclerView = recyclerView
    }

    fun getViewModel(viewModelClass: Class<VM>): VM? {
        if (setupViewModel() == null) throw ViewModelNotNullException()
        if (fragment != null) {
            return ViewModelProviders.of(fragment?.get()!!)
                    .get(viewModelClass)
        } else if (activity != null) {
            return ViewModelProviders.of(activity?.get()!!)
                    .get(viewModelClass)
        }
        return null
    }

    abstract fun setupViewModel(): Class<VM>?


    override
    fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH? {
        return null
    }

    override
    fun onBindViewHolder(holder: VH, position: Int) {
        if (listener != null) {
            holder.itemView.rootView.setOnClickListener(setOnClickItem(holder))
        }
    }

    private fun setOnClickItem(holder: VH): View.OnClickListener {
        return View.OnClickListener { v ->
            listener?.onClickHolder(
                    holder,
                    holder.adapterPosition )
        }
    }

}
